keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+The `metadata.name` field is optional as long as the config's GVK uniquely identifies it. In Kustomization, it is not required unless the config is being handled as a resource. In Composition, it is always recommended, because transformer config is always handled as a resource before execution and uniqueness must be global across all Compositions in the tree. 
+
+The `env` field for container plugins will be deprecated. In line with Kustomize's principles, plugin configuration should be passed in the KRM object and should not be affected by the execution environment. The network and storage mount options will be retained, as they are needed to support some common generator plugin use cases. Guidance on their appropriate use will be added to the plugin developer documentation.
+
+### Legacy plugin migration
+
+Although legacy plugins are still in alpha, the alpha has been around in standalone Kustomize for quite a while. Documentation on the migration process will be provided. See the [Rollout Plan](#rollout-plan) section for more detail on deprecation and removal of the existing alphas.
+
+The exec plugin conversion process should be relatively straightforward: copy the provider in-tree and update the configuration that referred to it to have a Provider stanza pointing to the new location. `argsFromFile` and `argsOneLiner` can be straightforwardly converted to the `args` subfield the KRM Functions style already supports.
+
+For authors who build plugins for distribution (as opposed to bespoke plugins where embedding within an end-user Kustomization is an appropriate solution), the recommendation will be to upgrade to the Catalog model. A Catalog guide will be added to the Kustomize website and will address the scenario of migrating a legacy style plugin.
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an end user, I want to develop a simple plugin for my particular Kustomization.
+
+1. Write a script that adheres to the [KRM Functions Specification] and place it within the Kustomization root.
+2. Write a corresponding plugin config and reference it from the Kustomization's generator or transformer field as appropriate.
+3. Run `kustomize edit generate-catalog`
+3. Run `kustomize build --trusted-plugin-catalog=catalog.yaml`
+
+```bash
+# Possible directory structure
+.
+⊢ Kustomization.yaml
+⊢ input.yaml
+⊢ reorder.yaml
+⊢ plugin-catalog.yaml
+∟ plugins
+  ∟ sorter
+    ⊢ sorter.rb
+    ∟ sorter_test.rb
+```
+
+```yaml
+# Example Kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- input.yaml
+transformers:
+- reorder.yaml
+```
+
+```yaml
+# Example reorder.yaml
+apiVersion: local-config.example.co/v1
```

Comment:
I think that making every step in the customization a fully qualified resource adds unnecessary burden on the kustomization author.  The necessary information here is the exec location.  I would consider a pipeline that's closer to what kpt has where it's focused on only the needed information.  Expanding to fully specified resources is an option but most of the plugins don't need that complexity, they mostly need the place for the executable code (binary, container registry, built-in) and parameters (positional, dictionary, rare cases it's a structure)

Reply:
I appreciate your desire to minimize what end users need to type, but this isn't complexity from Kustomize's POV: Kustomize is entirely [KRM-centric](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/resource-management.md), and intentionally exposes the KRM to end users. Kustomization itself is a KRM object, as are each of its transformers. From a declarative design perspective, the necessary information isn't the code to execute, but the declarative specification of the desired state, i.e. the KRM object. The apiVersion and kind are crucial to making that spec sustainable and scalable, and  can be used as a lookup key to find the code (analogous to a controller) that is capable of implementing the spec. That's how Catalog works; it radically reduces what end users need to type while retaining KRM orientation, and it removes implementation concerns from end-user configuration, in the case where those are different personas. Omitting the KRM is not an option for Kustomize, given the fundamental principles of its design.

Reply:
I don't think my position is to get rid of KRM in kustomize, far from it.  The question is at what level do we need to specify GVK.  The pipeline specification works just as well if the Kustomization is the KRM resource and the steps in the pipeline are just sub-objects.  This is how kpt pipeline is defined: https://kpt.dev/book/04-using-functions/01-declarative-function-execution.  The amount of boilerplate is shrunk tremendously.

Reply:
IIUC, the fundamental difference here is what part of the configuration we consider to be the boilerplate. In kpt, we see the GVK as boilerplate and hide that away by providing the `configMap` field. In kustomize, however, it seems that the community has established that the GVK is the important aspect of a plugin configuration, and the location of the implementation (the provider field here) is the additional boilerplate code that should be abstracted away from the kustomization file. 

The "sub objects" in the kustomization file - e.g. namePrefix, replacements, etc - also get converted under the hood to a builtin plugin configuration each with its own unique GVK. We could consider doing something analogous with function plugins, but the user would still have to provide a mapping from the function location to a unique GVK. 

Our kpt functions (to my knowledge) all accept `ConfigMap` as input, which is what allows us to remove the GVK from the Kptfile. IIRC we discussed out of band that we would require all function plugins to accept a unique GVK. If each function plugin needs to have a uniquely identifying GVK, is it still possible to remove that information from the function configuration?

Reply:
is there a way to automatically recognize something that has less boilerplate?  coming back to the nameprefix example.  I find that this is the ease people love about kustomize, would love to continue having raving fans.
```
namePrefix: staging-
commonLabels:
  variant: staging
  org: acmeCorporation
```

Reply:
Natasha's comment is spot-on: 

> IIUC, the fundamental difference here is what part of the configuration we consider to be the boilerplate. In kpt, we see the GVK as boilerplate and hide that away by providing the configMap field. In kustomize, however, it seems that the community has established that the GVK is the important aspect of a plugin configuration, and the location of the implementation (the provider field here) is the additional boilerplate that should be moved away from the kustomization file.

GVK is **not** boilerplate, it's the essence of the API. Those `namePrefix` and `commonLabels` fields are part of the schema of `kustomize.config.k8s.io/v1beta1/Kustomization`, and operations that are fundamental to Kustomize will continue to be exposed as top-level fields like that. Let's keep in mind that this KEP is about plugins, and nobody is suggesting that plugins replace core functionality in Kustomize. IMO the only existing built-in we'd extract is Helm, because we could make it work much better as a plugin than we can as a built-in, for technical reasons. It would be possible to retain the top-level fields for first-party code that is only a plugin for technical reasons as in that case. But that _cannot_ be true for third-party plugins, which cannot possibly have corresponding fields in Kustomization's schema. If an organization uses plugins so extensively that they have their own series of transformers they want to wrap up into simplified top-level fields a la Kustomization, they could totally do that by making their own meta plugin that does so.